File: src/Creational/Builder/InsurancePolicyDirector.cs
namespace DesignPatterns.Creational.Builder;

public class InsurancePolicyDirector
{
    public InsurancePolicy ConstructComprehensivePolicy(IInsurancePolicyBuilder builder)
    {
        return builder
            .WithPolicyType("Comprehensive")
            .WithPremium(1500.00)
            .WithCoverAmount(500000.00)
            .Build();
    }

    public InsurancePolicy ConstructBasicPolicy(IInsurancePolicyBuilder builder)
    {
        return builder
            .WithPolicyType("Basic")
            .WithPremium(500.00)
            .WithCoverAmount(100000.00)
            .Build();
    }
}

File: ts/src/creational/builder/InsurancePolicyDirector.ts
import { IInsurancePolicyBuilder } from './IInsurancePolicyBuilder';
import { InsurancePolicy } from './InsurancePolicy';

export class InsurancePolicyDirector {
    constructComprehensive(builder: IInsurancePolicyBuilder): InsurancePolicy {
        return builder
            .withPolicyType('Comprehensive')
            .withPremium(1500.0)
            .withCoverAmount(500000.0)
            .build();
    }

    constructBasic(builder: IInsurancePolicyBuilder): InsurancePolicy {
        return builder
            .withPolicyType('Basic')
            .withPremium(500.0)
            .withCoverAmount(100000.0)
            .build();
    }
}
